tensorboard/db_test.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import itertools
+import os
+import sqlite3
+
+import tensorflow as tf
+
+from tensorboard import db
+from tensorboard import test_util
+
+
+class DbTestCase(test_util.TestCase):
+
+  def __init__(self, *args, **kwargs):
+    super(DbTestCase, self).__init__(*args, **kwargs)
+    self._db_connection_provider = None
+
+  def setUp(self):
+    super(DbTestCase, self).setUp()
+    db_path = os.path.join(self.get_temp_dir(), 'DbTestCase.sqlite')
+    self._db_connection_provider = lambda: sqlite3.connect(db_path)
+    with contextlib.closing(self._db_connection_provider()) as conn:
+      with conn:
+        schema = db.Schema(conn)
+        schema.create_tables()
+        schema.create_indexes()
+
+
+class PluginTest(DbTestCase):
+
+  def testGetPluginIds(self):
+    tbase = db.TensorBase(self._db_connection_provider)
+    self.assertEqual({'b': 1}, tbase.get_plugin_ids(['b']))
+    self.assertEqual({'a': 2}, tbase.get_plugin_ids(['a']))
+    self.assertEqual({'b': 1}, tbase.get_plugin_ids(['b']))
+    self.assertEqual({'b': 1, 'c': 3}, tbase.get_plugin_ids(['c', 'b']))
+
+
+class IdTest(test_util.TestCase):
+
+  def testPositiveBits(self):
+    with self.assertRaises(ValueError):
+      db.Id('i', 0)
+    with self.assertRaises(ValueError):
+      db.Id('i', -1)
+
+  def testCheck(self):
+    id1 = db.Id('foo_id', 1)
+    with self.assertRaises(ValueError):
+      id1.check(-1)
+    id1.check(0)
+    id1.check(1)
+    with self.assertRaises(ValueError):
+      id1.check(2)
+    id2 = db.Id('foo_id', 2)
+    id2.check(0)
+    id2.check(3)
+    with self.assertRaises(ValueError):
+      id2.check(4)
+
+  def testGenerate(self):
+    self.assertIn(db.Id('foo_id', 1).generate(), (0, 1))
```

Comment:
Especially given that the RNG is not seeded, this test seems a bit weak…if you additionally run the check (a) with an `Id` whose `bits` is not `1` and (b) a thousand times, I'd be happier. :-)

(That said, the code is Clearly Correct…)